core/iwasm/libraries/debug-engine/debug_engine.c

```diff
@@ -0,0 +1,982 @@
+                            debug_object, APP_THREAD_STACK_SIZE_MAX)) {
+        goto fail2;
+    }
+
+    return control_thread;
+
+fail2:
+    os_cond_destroy(&control_thread->wait_cond);
+fail1:
+    os_mutex_destroy(&control_thread->wait_lock);
+fail:
+    wasm_runtime_free(control_thread);
+    return NULL;
+}
+
+static void
+wasm_debug_control_thread_destroy(WASMDebugObject *debug_object)
+{
+    WASMDebugControlThread *control_thread = debug_object->control_thread;
+    LOG_VERBOSE("control thread of debug objet %p stop at %s:%d\n",
+                debug_object, debug_object->control_thread->ip_addr,
+                debug_object->control_thread->port);
+    os_mutex_lock(&control_thread->wait_lock);
+    control_thread->status = STOPPED;
+    wasm_close_gdbserver(control_thread->server);
+    os_mutex_unlock(&control_thread->wait_lock);
+    os_cond_signal(&control_thread->wait_cond);
+    os_thread_join(control_thread->tid, NULL);
+    wasm_runtime_free(control_thread->server);
+
+    os_mutex_destroy(&control_thread->wait_lock);
+    os_cond_destroy(&control_thread->wait_cond);
+    wasm_runtime_free(control_thread);
+}
+
+static WASMDebugEngine *g_debug_engine;
+
+static WASMDebugEngine *
+wasm_debug_engin_create()
+{
+    WASMDebugEngine *engine;
+    if (!(engine = wasm_runtime_malloc(sizeof(WASMDebugEngine)))) {
+        LOG_ERROR("WASM Debug Engine error: failed to allocate memory");
+        return NULL;
+    }
+    memset(engine, 0, sizeof(WASMDebugEngine));
+
+    //TODO: support Wasm platform in LLDB
+    /* engine->control_thread = wasm_debug_control_thread_create((WASMDebugObject *)engine);
+    engine->control_thread->debug_engine = (WASMDebugObject *)engine;
+    engine->control_thread->debug_instance = NULL;
+    sprintf(engine->control_thread->ip_addr, "127.0.0.1");
+    engine->control_thread->port = 1234;*/
+
+    bh_list_init(&engine->debug_instance_list);
+
+    return engine;
+}
+
+bool
+wasm_debug_engin_init(char *ip_addr, int platform_port, int process_port)
+{
+    if (g_debug_engine == NULL)
+        g_debug_engine = wasm_debug_engin_create();
+
+    if (g_debug_engine) {
+        process_port -= 1;
+        g_debug_engine->platform_port =
+          platform_port > 0 ? platform_port : 1234;
+        g_debug_engine->process_base_port =
+          process_port > 0 ? process_port : 6169;
+        if (ip_addr)
+            sprintf(g_debug_engine->ip_addr, "%s", ip_addr);
+        else
+            sprintf(g_debug_engine->ip_addr, "%s", "127.0.0.1");
+        g_debug_engine->active = true;
+    }
+
+    return g_debug_engine != NULL;
+}
+
+void
+wasm_debug_set_engine_active(bool active)
+{
+    if (g_debug_engine) {
+        g_debug_engine->active = active;
+    }
+}
+
+bool
+wasm_debug_get_engine_active(void)
+{
+    if (g_debug_engine) {
+        return g_debug_engine->active;
+    }
+    return false;
+}
+
+void
+wasm_debug_engin_destroy()
+{
+    wasm_runtime_free(g_debug_engine);
+    g_debug_engine = NULL;
```

Comment:
If the `g_debug_engine` is not instantiated, then here we are freeing a NULL pointer

```suggestion
    if (g_debug_engine) {
        wasm_runtime_free(g_debug_engine);
        g_debug_engine = NULL;
    }
```